File: 03-11-listas/adts/list/linked_list.cpp
#pragma once

#include "./list.cpp"

template <class T>
class linked_list : public list<T>
{
private:
   struct node
   {
      T data;
      node *next;
   };

   node *head = nullptr;
   int length = 0;

public:
   linked_list() {}

   virtual void add(T data) override
   {
      if (this->head == nullptr)
      {
         this->head = new node{};
         this->head->data = data;
         this->head->next = nullptr;
      }
      else
      {
         // rellenar
      }
      this->length++;
   }

   virtual void remove(T data) override
   {
   }

   virtual void removeAt(int pos) override
   {
   }

   virtual bool exists(T data) override
   {
      node *aux = this->head;
      while (aux != nullptr)
      {
         if (aux->data == data)
         {
            return true;
         }

         aux = aux->next;
      }

      return false;
   }

   virtual int getPos(T data) override
   {
   }

   virtual T get(int pos) override
   {
   }

   virtual bool isEmpty() override
   {
      return this->length == 0;
   }

   virtual int size() override
   {
      return this->length;
   }
};

File: iterator/adts/list/LinkedList.java
package adts.list;

import java.util.Iterator;

public class LinkedList<T> implements List<T> {
   private class node {
      T data;
      node next;
   }

   private node head = null;
   private int length = 0;

   private class LinkedListIterator implements Iterator<T> {
      node curr;

      LinkedListIterator(node head) {
         this.curr = head;
      }

      @Override
      public boolean hasNext() {
         return this.curr != null;
      }

      @Override
      public T next() {
         if (this.curr == null) {
            throw new RuntimeException("iterator has reached the end");
         }
         T ret = this.curr.data;
         this.curr = this.curr.next;
         return ret;
      }
   }

   public LinkedList() {
   }

   @Override
   public void add(T data) {
      if (this.head == null) {
         this.head = new node();
         this.head.data = data;
         this.head.next = null;
      } else {
         // ESTO ESTA MAL NO SE AGREGA AL PPIO. SE TIENE QUE AGREGAR AL FINAL
         node newN = new node();
         newN.data = data;
         newN.next = this.head;
         this.head = newN;
      }
      this.length++;
   }

   @Override
   public void remove(T data) {
      // TODO Auto-generated method stub
      throw new UnsupportedOperationException("Unimplemented method 'remove'");
   }

   @Override
   public void removeAt(int pos) {
      // TODO Auto-generated method stub
      throw new UnsupportedOperationException("Unimplemented method 'removeAt'");
   }

   @Override
   public boolean exists(T data) {
      node aux = this.head;
      while (aux != null) {
         if (aux.data == data) {
            return true;
         }

         aux = aux.next;
      }

      return false;
   }

   @Override
   public int getPos(T data) {
      // TODO Auto-generated method stub
      throw new UnsupportedOperationException("Unimplemented method 'getPos'");
   }

   @Override
   public T get(int pos) {
      // TODO Auto-generated method stub
      throw new UnsupportedOperationException("Unimplemented method 'get'");
   }

   @Override
   public boolean isEmpty() {
      // TODO Auto-generated method stub
      throw new UnsupportedOperationException("Unimplemented method 'isEmpty'");
   }

   @Override
   public int size() {
      // TODO Auto-generated method stub
      throw new UnsupportedOperationException("Unimplemented method 'size'");
   }

   @Override
   public Iterator<T> iterator() {
      return new LinkedListIterator(this.head);
   }

}
